securedrop/passphrases.py

```diff
@@ -0,0 +1,123 @@
+
+    _PASSPHRASE_WORDS_COUNT = 7
+
+    # Enforce a reasonable maximum length for passphrases to avoid DoS
+    _MAX_PASSPHRASE_LENGTH = 128
+    _MIN_PASSPHRASE_LENGTH = 20
+
+    def __init__(
+        self, language_to_words: Dict[str, List[str]], fallback_language: str = "en"
+    ) -> None:
+        # SystemRandom sources from the system rand (e.g. urandom, CryptGenRandom, etc)
+        # It supplies a CSPRNG but with an interface that supports methods like choice
+        self._random_generator = SystemRandom()
+
+        self._fallback_language = fallback_language
+        self._language_to_words = language_to_words
+        if self._fallback_language not in self._language_to_words:
+            raise InvalidWordsListError(
+                "Missing words list for fallback language {}".format(self._fallback_language)
+            )
+
+        # Validate each words list
+        for language, words_list in self._language_to_words.items():
+            # Ensure all words are ascii
+            try:
+                " ".join(words_list).encode("ascii")
+            except UnicodeEncodeError:
+                raise InvalidWordsListError("A words list contains non-ASCII words.")
+
+            # Ensure that passphrases longer than what's supported can't be generated
+            longest_word = max(words_list, key=len)
+            longest_passphrase_length = len(longest_word) * self._PASSPHRASE_WORDS_COUNT
+            longest_passphrase_length += self._PASSPHRASE_WORDS_COUNT  # One space between each word
+            if longest_passphrase_length >= self._MAX_PASSPHRASE_LENGTH:
+                raise InvalidWordsListError(
+                    "Passphrases over the maximum length ({}) may be generated:"
+                    " longest word in words list for language {} is '{}' and number of words per"
+                    " passphrase is {}".format(
+                        self._MAX_PASSPHRASE_LENGTH,
+                        language,
+                        longest_word,
+                        self._PASSPHRASE_WORDS_COUNT,
+                    )
+                )
+
+            # Ensure that passphrases shorter than what's supported can't be generated
+            shortest_word = min(words_list, key=len)
+            shortest_passphrase_length = len(shortest_word) * self._PASSPHRASE_WORDS_COUNT
+            shortest_passphrase_length += self._PASSPHRASE_WORDS_COUNT
+            if shortest_passphrase_length <= self._MIN_PASSPHRASE_LENGTH:
+                raise InvalidWordsListError(
+                    "Passphrases under the minimum length ({}) may be generated:"
+                    " shortest word in words list for language {} is '{}' and number of words per"
+                    " passphrase is {}".format(
+                        self._MIN_PASSPHRASE_LENGTH,
+                        language,
+                        shortest_word,
+                        self._PASSPHRASE_WORDS_COUNT,
+                    )
+                )
```

Comment:
For this purpose, I like the approach of having "clean"/reliable word list, instead of having to do more work at runtime to clean them up ie. try several passphrase until we find a good one. This adds complexity, requires dedicated test cases (a 100% bad words list, a semi-bad words list, etc.), will lead to bugs and surprises such as passphrase generation taking longer than expected, etc.. 

Creating clean word lists seems like a quick, one-time effort. Maintaining code that can handle bad word lists will take more effort/work (over time), I think,

Reply:
Another option would be to just filter out bad words from the supplied lists when initializing the Passphrase generator - this is still somewhat simple and doesn't require "perfect" words list.

Reply:
All right. We can revisit this when/if we add better checks and more locale-specific lists than we have now.

Reply:
So I ended up implementing what I described in my last comment as I think it's a good improvement over my previous approach:

* The code will now take any word list, but will ignore words that are less than 4 characters. 
* Additionally, it will raise an `InvalidWordListError` if a word list has less than 6000 "long-enough" words, to avoid the case where a word list has like 5 long-enough words so the passphrase are totally not random/secure.
* As before, it will also raise an error if passphrases that are too short may be generated.

This allowed me to re-enable the french word list.

Reply:
See the comment above about Diceware entropy.